### PR TITLE
8.0 4171 colli info per carrier

### PR DESCRIPTION
**Pull request by daramousk**
Refactors synchronization code and allows the user to set a different packaging type per carrier

1) Make sure that when the sync is done, server wins (local data will be overwritten)
2) A new table has been created that holds the package types of transsmart
3) A new field has been inserted on res_partner that you can use to set the default package type for that carrier
4) You must set a a default package type on the carriers or on the default carrier, if not the sync will fail with a rather non descriptive error message that will be fixed on a later pull request

@hbrunn @gcapalbo @NL66278 Please review this

### CODE REVIEW

File: delivery_transsmart/models/res_config.py [hunk 1]
```diff
@@ -108,47 +108,87 @@ def get_transsmart_service(self):
...
+        """
+        assembling an odata $filter
+        http://www.odata.org/documentation/
+        odata-version-2-0/uri-conventions/
+
+        Gets all the records that their Id is not in transsmart_ids
+        """
+        _filter = ''
+        for index, code in enumerate(transsmart_ids):
+            if index < len(transsmart_ids) - 1:
```
Comment (NL66278) [inline delivery_transsmart/models/res_config.py]: The if and the else logic are nearly the same,  except for the word and.  In these cases have the if else conditions fill a variable with ' and '  or '' and then make the rest one statement.

File: delivery_transsmart/models/stock_picking.py [hunk 2]
```diff
@@ -99,8 +99,17 @@ def get_invoice_name(self):
...
-
+        #weight = float(round(((self.weight * 1.03) + 0.05), 1))
+        # the code below assumes that trassmart_id is unique...
+        carrier = self.env['res.partner'].search(
+                [('transsmart_id', '=', self.get_carrier_id())])
+        colli_information = carrier.transsmart_package_type_id
+        if not colli_information:
+            raise Warning(_(
+                "You have not set a package type for"
+                " the carrier {}, please do".format(carrier.name) +
```
Comment (NL66278) [inline delivery_transsmart/models/stock_picking.py]: Here you have a variable part in a translatable string. That is something to avoid.

File: delivery_transsmart/models/stock_picking.py [hunk 3]
```diff
@@ -99,8 +99,17 @@ def get_invoice_name(self):
     def _transsmart_document_from_stock_picking(self):
         # TODO: Move to transsmart settings
```
Comment (NL66278) [inline delivery_transsmart/models/stock_picking.py:100]: Maybe a good idea to do the TODO?

File: delivery_transsmart/models/stock_picking.py [hunk 4]
```diff
@@ -131,13 +140,13 @@ def _transsmart_document_from_stock_picking(self):
             "AddressCountry": self.partner_id.country_id.code or '',
             "ColliInformation": [
```
Comment (NL66278) [inline delivery_transsmart/models/stock_picking.py]: Just collapse the [ and { together, as wll as the } and ], and you can save two lines and an indentation level

File: delivery_transsmart/models/transsmart_package_type.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+# © 2017 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openerp import fields, models
+
+
+class TranssmartPackageType(models.Model):
+    _name = 'transsmart.package.type'
+
+    name = fields.Char()
+    _type = fields.Char()
```
Comment (NL66278) [inline delivery_transsmart/models/transsmart_package_type.py]: Do not start a field name with _. It suggest it is something private, while I think you just want to avoid a warning on using a reserved word as variable/field name. Call it something like package_type.

File: delivery_transsmart/models/res_config.py [hunk 6]
```diff
@@ -108,47 +108,87 @@ def get_transsmart_service(self):
     def get_transsmart_carrier_tag(self):
         return self.env['ir.model.data'].get_object('delivery_transsmart', 'res_partner_category_transsmart_carrier')
```
Comment (NL66278) [inline delivery_transsmart/models/res_config.py]: Not changed by you, but might better use ref() to get a record through ir.model.data.

File: delivery_transsmart/models/res_config.py [hunk 7]
```diff
@@ -108,47 +108,87 @@ def get_transsmart_service(self):
...
-        local_codes = {local.code: local for local in local_data}
+        local_transsmart_ids = [local.transsmart_id for local in local_data]
+        params = {'$filter': self._get_odata_filter(local_transsmart_ids)}
+        remote_data = self.get_transsmart_service().receive(
+            '/ServiceLevelOther',
+            params=params if not update_local_data else {})
         for data in remote_data:
-            if not data['Code'] in local_codes:
+            if not data['Id'] in local_transsmart_ids:
                 self.env['delivery.service.level'].create({
```
Comment (NL66278) [inline delivery_transsmart/models/res_config.py]: Put self.env stuff outside loop, use a service_level_model as variable

File: delivery_transsmart/models/res_config.py [hunk 8]
```diff
@@ -108,47 +108,87 @@ def get_transsmart_service(self):
...
                     'code': data['Code'],
```
Comment (NL66278) [inline delivery_transsmart/models/res_config.py]: And in the same spirit of DRY: use a vals dictionary for data to be used on eiither create or write.

File: delivery_transsmart/models/res_config.py [hunk 9]
```diff
@@ -108,47 +108,87 @@ def get_transsmart_service(self):
...
+                _logger.info("Updated Service Level {}".format(
+                    rec_to_be_updated.transsmart_id))
 
-        remote_data = self.get_transsmart_service().receive('/ServiceLevelTime')
         local_data = self.env['delivery.service.level.time'].search([])
-        local_codes = {local.code: local for local in local_data}
+        local_transsmart_ids = [local.transsmart_id for local in local_data]
+        params = {'$filter': self._get_odata_filter(local_transsmart_ids)}
+        remote_data = self.get_transsmart_service().receive(
+                '/ServiceLevelTime',
```
Comment (NL66278) [inline delivery_transsmart/models/res_config.py]: Too much indent

File: delivery_transsmart/models/res_config.py [hunk 10]
```diff
@@ -108,47 +108,87 @@ def get_transsmart_service(self):
...
+        _filter = ''
+        for index, code in enumerate(transsmart_ids):
+            if index < len(transsmart_ids) - 1:
+                _filter += 'Id ne {} and '.format(transsmart_ids[index])
+            else:
+                _filter += 'Id ne {}'.format(transsmart_ids[index])
+        return _filter
+
     @api.multi
     def sync_transsmart_models(self):
```
Comment (NL66278) [inline delivery_transsmart/models/res_config.py]: I see a lot of nearly identical code repeating. Like get transsmart data for some model. Filter it. Loop over it. Write what we already have, create what we do not have. It should really be possible to make some general method and call it with the arguments (or maybe a dictionary of info) to handle what is different in those loops.

File: delivery_transsmart/models/res_config.py (hunk 6 above)
Comment (gfcapalbo) [reply delivery_transsmart/models/res_config.py]: use ref(), it is a more compact sig. even if you did not change this.

File: delivery_transsmart/models/res_config.py (hunk 7 above)
Comment (gfcapalbo) [reply delivery_transsmart/models/res_config.py]: get the model outside of loop. append   _model to your name to signify it is self.env['$modelname']

File: delivery_transsmart/models/res_config.py [hunk 11]
```diff
@@ -108,47 +108,87 @@ def get_transsmart_service(self):
...
             else:
-                local_codes[data['Code']].write({
+                rec_to_be_updated = local_data.filtered(
+                        lambda rec: rec.transsmart_id == data['Id'])
+                rec_to_be_updated.write({
                     'code': data['Code'], 
                     'name': data['Name'], 
                     'transsmart_id': data['Id']})
+                _logger.info("Updated Service Level {}".format(
+                    rec_to_be_updated.transsmart_id))
```
Comment (gfcapalbo) [inline delivery_transsmart/models/res_config.py]: I prefer to close the ) on another line. Bothe are flake8 compliant, and my habit does make code longer , but i think more readable nested functions.

File: delivery_transsmart/models/res_config.py (hunk 9 above)
Comment (gfcapalbo) [reply delivery_transsmart/models/res_config.py]: indent

File: delivery_transsmart/models/res_config.py [hunk 12]
```diff
@@ -157,32 +197,78 @@ def sync_transsmart_models(self):
...
+                rec_to_be_updated.write({
+                    'code': data['Code'],
+                    'name': data['Name'],
                     'transsmart_id': data['Id']})
-
+                _logger.info("Updated transsmart.cost.center {}".format(
+                    rec_to_be_updated.transsmart_id))
+        # get the packages (box, pallet etc...)
+        local_data = self.env['transsmart.package.type'].search([])
+        local_transsmart_ids = [local.transsmart_id for local in local_data]
```
Comment (gfcapalbo) [inline delivery_transsmart/models/res_config.py]: this  is nice but can be achieved by mapping your search 
self.env['transsmart.package.type'].search([]).mapped('transsmart_id')
one less line of code, same complexity.

File: delivery_transsmart/models/res_config.py [hunk 13]
```diff
@@ -157,32 +197,78 @@ def sync_transsmart_models(self):
...
                     'is_company': True, 
                     'transsmart_id': data['Id'],
                     'category_id': [(4,self.get_transsmart_carrier_tag().id)]})
+                _logger.info("Updated res.partner {}".format(
+                    rec_to_be_updated.transsmart_id))
 
-        remote_data = self.get_transsmart_service().receive('/Costcenter')
         local_data = self.env['transsmart.cost.center'].search([])
-        local_codes = {local.code: local for local in local_data}
+        local_transsmart_ids = [local.transsmart_id for local in local_data]
```
Comment (gfcapalbo) [inline delivery_transsmart/models/res_config.py]: you can map your search :  search([]).mapped('transsmart_id')

File: delivery_transsmart/models/res_partner.py [hunk 14]
```diff
@@ -33,7 +33,9 @@ class ResPartner(models.Model):
         string="Transsmart Code")
     transsmart_id = fields.Integer(
         "Transsmart ID")
```
Comment (gfcapalbo) [inline delivery_transsmart/models/res_partner.py:35]: use one line when you can.

File: delivery_transsmart/models/transsmart_package_type.py (hunk 5 above)
Comment (gfcapalbo) [inline delivery_transsmart/models/transsmart_package_type.py]: mm not standard naming, avoid type  (using keywords maked confusing code) extend it with a more verbose naming.

File: delivery_transsmart/models/res_config.py (hunk 7 above)
Comment (gfcapalbo) [inline delivery_transsmart/models/res_config.py]: Flake8 and pylint the whole code if you have time (long lines and unused imports) make another MR for that, so the work you did here is not confused or less visible do to  the many flake/lint changes.

File: delivery_transsmart/models/res_config.py [hunk 15]
```diff
@@ -108,47 +108,87 @@ def get_transsmart_service(self):
...
+        odata-version-2-0/uri-conventions/
+
+        Gets all the records that their Id is not in transsmart_ids
+        """
+        _filter = ''
+        for index, code in enumerate(transsmart_ids):
+            if index < len(transsmart_ids) - 1:
+                _filter += 'Id ne {} and '.format(transsmart_ids[index])
+            else:
+                _filter += 'Id ne {}'.format(transsmart_ids[index])
```
Comment (gfcapalbo) [inline delivery_transsmart/models/res_config.py]: this can be compacted only  'and' is the difference.  
_filter += 'Id ne {} {}'.format(transsmart_ids[index],  'and' if index < len(transsmart_ids) - 1 else '')

and remove all the if/else , save 4 lines of code,  i think more readable?

File: delivery_transsmart/models/res_config.py [hunk 16]
```diff
@@ -108,47 +108,87 @@ def get_transsmart_service(self):
...
-                      "this warning from the code in the Transsmart Delivery module")
-        remote_data = self.get_transsmart_service().receive('/ServiceLevelOther')
+        # if true, fetch everything from the server and overwrite the existing
+        # values of the local records
+        # if false, set up a filter on our request and don't fetch stuff we
+        # already have.
+        update_local_data = True
         local_data = self.env['delivery.service.level'].search([])
-        local_codes = {local.code: local for local in local_data}
+        local_transsmart_ids = [local.transsmart_id for local in local_data]
```
Comment (gfcapalbo) [inline delivery_transsmart/models/res_config.py]: map!  search([]),mapped('transsmart_id')

File: delivery_transsmart/models/res_config.py [hunk 17]
```diff
@@ -108,47 +108,87 @@ def get_transsmart_service(self):
...
+        params = {'$filter': self._get_odata_filter(local_transsmart_ids)}
+        remote_data = self.get_transsmart_service().receive(
+            '/ServiceLevelOther',
+            params=params if not update_local_data else {})
```
Comment (gfcapalbo) [inline delivery_transsmart/models/res_config.py]: clearer if closed on another line, just my preference, if you prefer this , it is flake/lint compliant, please keep on like this.

File: delivery_transsmart/models/res_config.py [hunk 18]
```diff
@@ -106,7 +106,10 @@ def get_transsmart_service(self):
         return self.env['delivery.web.service'].browse([wst])
 
     def get_transsmart_carrier_tag(self):
-        return self.env['ir.model.data'].get_object('delivery_transsmart', 'res_partner_category_transsmart_carrier')        
+        #return self.env['ir.model.data'].get_object('delivery_transsmart', 'res_partner_category_transsmart_carrier')
```
Comment (NL66278) [inline delivery_transsmart/models/res_config.py]: Do not leave the commented stuff. We have source management system to see what was there before

Comment (daramousk) [reply delivery_transsmart/models/res_config.py]: @NL66278 Yes, missed this one, now removed

File: delivery_transsmart/models/res_config.py [hunk 19]
```diff
@@ -106,83 +106,150 @@ def get_transsmart_service(self):
         return self.env['delivery.web.service'].browse([wst])
 
     def get_transsmart_carrier_tag(self):
-        return self.env['ir.model.data'].get_object('delivery_transsmart', 'res_partner_category_transsmart_carrier')        
+        return self.env.ref(
+                'delivery_transsmart.res_partner_category_transsmart_carrier'
```
Comment (NL66278) [inline delivery_transsmart/models/res_config.py]: Fix indent. Last ).id can go on line before, or if you prefer to keep it on its own line, the ) should be in the same column as the r of return.

File: delivery_transsmart/models/res_config.py [hunk 20]
```diff
@@ -106,83 +106,150 @@ def get_transsmart_service(self):
...
+                ).id
+
+    def _get_odata_filter(self, transsmart_ids):
```
Comment (NL66278) [inline delivery_transsmart/models/res_config.py]: Not used, so delete

File: delivery_transsmart/views/res_partner.xml [hunk 21]
```diff
@@ -0,0 +1,16 @@
+<openerp>
+    <data>
+
+        <record id="form_res_partner_carrier_package" model="ir.ui.view">
+            <field name="name">Carrier package field</field>
+            <field name="model">res.partner</field>
+            <field name="inherit_id" ref="base.view_partner_form" />
+            <field name="arch" type="xml">
+                <xpath expr="//field[@name='date']" position="after">
```
Comment (NL66278) [inline delivery_transsmart/views/res_partner.xml:9]: Not interesting for most partners. Move to some appropiate tab.

File: delivery_transsmart/models/transsmart_package_type.py [hunk 22]
```diff
@@ -0,0 +1,17 @@
+# -*- coding: utf-8 -*-
+# © 2017 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openerp import fields, models
+
+
+class TranssmartPackageType(models.Model):
+    _name = 'transsmart.package.type'
+
```
Comment (NL66278) [inline delivery_transsmart/models/transsmart_package_type.py]: model missing _description

File: delivery_transsmart/models/res_partner.py [hunk 23]
```diff
@@ -33,7 +33,9 @@ class ResPartner(models.Model):
...
-
+    transsmart_package_type_id = fields.Many2one(
+            'transsmart.package.type',
```
Comment (NL66278) [inline delivery_transsmart/models/res_partner.py]: indent

**Comment by daramousk** [comment]
@NL66278 Fully refactored again, cut down to 1/3 of its size

**Comment by lfreeke** [comment]
2017-10-24 12:49:10,765 7019 ERROR ari_test openerp.sql_db: bad query: ALTER TABLE "res_partner" ADD FOREIGN KEY ("transsmart_package_type_id") REFERENCES "delivery_package_type" ON DELETE set null
Traceback (most recent call last):
  File "/home/odoariacp80/odoo/parts/odoo/openerp/sql_db.py", line 247, in execute
    res = self._obj.execute(query, params)
IntegrityError: insert or update on table "res_partner" violates foreign key constraint "res_partner_transsmart_package_type_id_fkey"
DETAIL:  Key (transsmart_package_type_id)=(2) is not present in table "delivery_package_type".

Traceback (most recent call last):
  File "./bin/upgrade_odoo", line 79, in <module>
    sys.exit(anybox.recipe.odoo.runtime.upgrade.upgrade('/home/odoariacp80/odoo/upgrade.py', 'run', '/home/odoariacp80/odoo/etc/odoo.cfg', '/home/odoariacp80/odoo'))
  File "/home/odoariacp80/odoo/eggs/anybox.recipe.odoo-1.9.3.dev0-py2.7.egg/anybox/recipe/odoo/runtime/upgrade.py", line 151, in upgrade
    statuscode = getattr(upgrade_module, upgrade_callable)(session, logger)
  File "/home/odoariacp80/odoo/upgrade.py", line 19, in run
    session.update_modules(['all'])
  File "/home/odoariacp80/odoo/eggs/anybox.recipe.odoo-1.9.3.dev0-py2.7.egg/anybox/recipe/odoo/runtime/session.py", line 418, in update_modules
    db, update_module=True)
  File "/home/odoariacp80/odoo/parts/odoo/openerp/modules/registry.py", line 339, in get
    update_module)
  File "/home/odoariacp80/odoo/parts/odoo/openerp/modules/registry.py", line 370, in new
    openerp.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/home/odoariacp80/odoo/parts/odoo/openerp/modules/loading.py", line 350, in load_modules
    force, status, report, loaded_modules, update_module)
  File "/home/odoariacp80/odoo/parts/odoo/openerp/modules/loading.py", line 255, in load_marked_modules
    loaded, processed = load_module_graph(cr, graph, progressdict, report=report, skip_modules=loaded_modules, perform_checks=perform_checks)
  File "/home/odoariacp80/odoo/parts/odoo/openerp/modules/loading.py", line 157, in load_module_graph
    init_module_models(cr, package.name, models)
  File "/home/odoariacp80/odoo/parts/odoo/openerp/modules/module.py", line 297, in init_module_models
    obj._auto_end(cr, {'module': module_name})
  File "/home/odoariacp80/odoo/parts/pos/pos_remove_pos_category/product.py", line 106, in _auto_end
    return _auto_end_original(self, cr, context=context)
  File "/home/odoariacp80/odoo/parts/odoo/openerp/api.py", line 268, in wrapper
    return old_api(self, *args, **kwargs)
  File "/home/odoariacp80/odoo/parts/odoo/openerp/models.py", line 2733, in _auto_end
    cr.execute('ALTER TABLE "%s" ADD FOREIGN KEY ("%s") REFERENCES "%s" ON DELETE %s' % (t, k, r, d))
  File "/home/odoariacp80/odoo/parts/odoo/openerp/sql_db.py", line 171, in wrapper
    return f(self, *args, **kwargs)
  File "/home/odoariacp80/odoo/parts/odoo/openerp/sql_db.py", line 247, in execute
    res = self._obj.execute(query, params)
psycopg2.IntegrityError: insert or update on table "res_partner" violates foreign key constraint "res_partner_transsmart_package_type_id_fkey"
DETAIL:  Key (transsmart_package_type_id)=(2) is not present in table "delivery_package_type".
